File: app/src/test/java/com/example/adiphong/rx2/MapTest.java
package com.example.adiphong.rx2;

import org.junit.Test;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableEmitter;
import io.reactivex.FlowableOnSubscribe;
import io.reactivex.Maybe;
import io.reactivex.MaybeEmitter;
import io.reactivex.MaybeOnSubscribe;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Single;
import io.reactivex.SingleEmitter;
import io.reactivex.SingleOnSubscribe;


/**
 * Created by adiphong on 28/11/2017 AD.
 */

public class MapTest {

    //just
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    public void flowableJust() {
        Flowable.just("Hello", "World")
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------flowable  result " + s)
                );
    }

    @Test
    public void observableJust() {
        Observable.just("Hello", "World")
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------observable result " + s)
                );
    }

    @Test
    public void maybeJust() {
        // can be only one data
        Maybe.just("Hello")
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------maybe result " + s)
                );
    }

    @Test
    public void singleJust() {
        // can be only one data
        Single.just("Hello")
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------single result " + s)
                );
    }


    //create
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    public void flowableCreate() {
        Flowable.create(new FlowableOnSubscribe<String>() {
            @Override
            public void subscribe(FlowableEmitter<String> e) throws Exception {
                if (!e.isCancelled()) {
                    e.onNext("hello");
                    e.onNext("world");
                    e.onComplete();
                }
            }
        }, BackpressureStrategy.BUFFER)
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------flowable  result " + s)
                );
    }

    @Test
    public void observableCreate() {
        Observable.create(new ObservableOnSubscribe<String>() {
            @Override
            public void subscribe(ObservableEmitter<String> e) throws Exception {
                if (!e.isDisposed()) {
                    e.onNext("hello");
                    e.onNext("world");
                    e.onComplete();
                }
            }
        })
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------observable result " + s)
                );
    }

    @Test
    public void maybeCreate() {
        // can be only one data
        Maybe.create(new MaybeOnSubscribe<String>() {
            @Override
            public void subscribe(MaybeEmitter<String> e) throws Exception {
                if (!e.isDisposed()) {
                    e.onSuccess("hello");
                    e.onSuccess("world");
                    e.onComplete();
                }

            }
        }).map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------maybe result " + s)
                );
    }

    @Test
    public void singleCreate() {
        // can be only one data
        Single.create(new SingleOnSubscribe<String>() {
            @Override
            public void subscribe(SingleEmitter<String> e) throws Exception {
                if (!e.isDisposed()) {
                    e.onSuccess("hello");
                    e.onSuccess("world");
                }

            }
        }).map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------single result " + s)
                );
    }


    //fromCallAble
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    public void flowableFromCallable() {
        Flowable.fromCallable(() -> "hello")
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------flowable  result " + s)
                );
    }

    @Test
    public void observableFromCallable() {
        Observable.fromCallable(() -> "hello")
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------observable result " + s)
                );
    }

    @Test
    public void maybeFromCallable() {
        // can be only one data
        Maybe.fromCallable(() -> "hello")
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------maybe result " + s)
                );
    }

    @Test
    public void singleFromCallable() {
        // can be only one data
        Single.fromCallable(() -> "hello")
                .map(s -> s.length())
                .subscribe(
                        s -> System.out.println("---------------single result " + s)
                );
    }

}

File: app/src/test/java/com/example/adiphong/rx2/MapTest.kt
package com.example.adiphong.rx2

import io.reactivex.*
import org.junit.Test

/**
 * Created by adiphong on 28/11/2017 AD.
 */

class MapTest {

    //just
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    fun flowableJust() {
        Flowable.just(
                "Hello", "World"
        ).map { s ->
            s.length
        }.subscribe { s ->
            println("---------------flowable  result " + s!!)
        }
    }

    @Test
    fun observableJust() {
        Observable.just(
                "Hello", "World"
        ).map { s ->
            s.length
        }.subscribe { s ->
            println("---------------observable result " + s!!)
        }
    }

    @Test
    fun maybeJust() {
        // can be only one data
        Maybe.just(
                "Hello"
        ).map { s ->
            s.length
        }.subscribe { s ->
            println("---------------maybe result " + s!!)
        }
    }

    @Test
    fun singleJust() {
        // can be only one data
        Single.just(
                "Hello"
        ).map { s ->
            s.length
        }.subscribe { s ->
            println("---------------single result " + s!!)
        }
    }


    //create
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    fun flowableCreate() {
        Flowable.create(FlowableOnSubscribe<String> { e ->
            if (!e.isCancelled) {
                e.onNext("hello")
                e.onNext("world")
                e.onComplete()
            }
        },
                BackpressureStrategy.BUFFER
        ).map { s ->
            s.length
        }.subscribe { s ->
            println("---------------flowable  result " + s!!)
        }
    }

    @Test
    fun observableCreate() {
        Observable.create(ObservableOnSubscribe<String> { e ->
            if (!e.isDisposed) {
                e.onNext("hello")
                e.onNext("world")
                e.onComplete()
            }
        }).map { s ->
            s.length
        }.subscribe { s ->
            println("---------------observable result " + s!!)
        }
    }

    @Test
    fun maybeCreate() {
        // can be only one data
        Maybe.create(MaybeOnSubscribe<String> { e ->
            if (!e.isDisposed) {
                e.onSuccess("hello")
                e.onSuccess("world")
                e.onComplete()
            }
        }).map { s ->
            s.length
        }.subscribe { s ->
            println("---------------maybe result " + s!!)
        }
    }

    @Test
    fun singleCreate() {
        // can be only one data
        Single.create(SingleOnSubscribe<String> { e ->
            if (!e.isDisposed) {
                e.onSuccess("hello")
                e.onSuccess("world")
            }
        }).map { s ->
            s.length
        }.subscribe { s ->
            println("---------------single result " + s!!)
        }
    }


    //fromCallAble
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    fun flowableFromCallable() {
        Flowable.fromCallable { "hello" }
                .map { s -> s.length }
                .subscribe { s -> println("---------------flowable  result " + s!!) }
    }

    @Test
    fun observableFromCallable() {
        Observable.fromCallable { "hello" }
                .map { s -> s.length }
                .subscribe { s -> println("---------------observable result " + s!!) }
    }

    @Test
    fun maybeFromCallable() {
        // can be only one data
        Maybe.fromCallable { "hello" }
                .map { s -> s.length }
                .subscribe { s -> println("---------------maybe result " + s!!) }
    }

    @Test
    fun singleFromCallable() {
        // can be only one data
        Single.fromCallable { "hello" }
                .map { s -> s.length }
                .subscribe { s -> println("---------------single result " + s!!) }
    }

}
